plugins/chain_plugin/get_info_db.cpp

```diff
@@ -0,0 +1,159 @@
+         , server_version_string(app().version_string())
+         , server_full_version_string(app().full_version_string()) {}
+
+      // Called on accepted_block signal.
+      void on_accepted_block() {
+         try {
+            if (!get_info_enabled) {
+               return;
+            }
+
+            store_info();
+         } FC_LOG_AND_DROP(("get_info_db_impl on_accepted_block ERROR"));
+      }
+
+      // Called on irreversible_block signal.
+      void on_irreversible_block(const chain::signed_block_ptr& block, const block_id_type& id) {
+         try {
+            if (!get_info_enabled) {
+               return;
+            }
+
+            store_info(block, id);
+         } FC_LOG_AND_DROP(("get_info_db_impl on_irreversible_block ERROR"));
+      }
+
+      // Returns cached get_info results
+      get_info_db::get_info_results get_info() {
+         // safely load the info_cache pointer
+         std::shared_ptr<get_info_db::get_info_results> info = std::atomic_load(&info_cache);
+
+         if (info && info->contains_full_data()) {
+            return *info;
+         }
+
+         // This only happens right after initialization when starts from
+         // snapshot as no signals are emitted. We need to cache the current states.
+         store_info();
+
+         info = std::atomic_load(&info_cache);
+         assert(info);
+         return *info;
+      }
+
+   private:
+      // A handle to the controller.
+      const chain::controller& controller;
+
+      // Indication whether get_info RPC is enabled.
+      const bool get_info_enabled = false;
+
+      // Cache to store the current get_info results.
+      // Lock free by using std::atomic_load and std::atomic_store.
+      std::shared_ptr<get_info_db::get_info_results> info_cache = nullptr;
+
+      // Fixed data
+      std::string           server_version;
+      chain::chain_id_type  chain_id;
+      std::string           server_version_string;
+      std::string           server_full_version_string;
+
+      void store_info_common(const std::shared_ptr<get_info_db::get_info_results>& info) {
+         assert(info);
+
+         // fixed part
+         info->server_version             = server_version;
+         info->chain_id                   = chain_id;
+         info->server_version_string      = server_version_string;
+         info->server_full_version_string = server_full_version_string;
+
+         // chain head part
+         const auto& head = controller.head();
+         if (head.is_valid()) {
+            info->head_block_id       = head.id();
+            info->head_block_num      = block_header::num_from_id(info->head_block_id);
+            info->head_block_time     = head.block_time();
+            info->head_block_producer = head.producer();
+         }
+
+         // fork_db part
+         if (controller.fork_db_has_root()) {
+            info->fork_db_head_block_id        = controller.fork_db_head().id();
```

Comment:
You can use `fork_db_head()` and then check if valid to avoid an extra mutex lock in fork db.

Reply:
`earliest_available_block_num()` requires `fork_db_root_block_num()` which in turn requires fork_db_has_root. Checking if the returned head of fork_db_head() is valid does not reduce the total number of locks in this code section.

Reply:
`fork_db_head()` returning a valid `block_handle` means `fork_db_root` is available.

Reply:
https://github.com/AntelopeIO/spring/pull/920/commits/20e31401e6cf5664f6e55df94d212768202918ae